### PR TITLE
refactor(amazon/loadBalancer): Migrate LoadBalancerLocation and SecurityGroups to RxJS streams

### PR DESCRIPTION
**Pull request by christopherthielen**
This refactor fixes some issues with data not updating in the Load Balancer Dialog when fields are updated.    Note: This PR also includes the commit from https://github.com/spinnaker/deck/pull/6316, which should be merged first.

I've been playing with using RxJS to manage reactivity in formik forms and I think I am happy where this refactor ended up.  This refactor bridges component local state, formik state, and uses RxJS to do it.   I am interested to get feedback from others if this is a reasonable approach or not.  Primarily review the `LoadBalancerLocation` component which is where most of the reactive logic lives. 

### Here are some pros:

1) Responding to changes inside Formik is reasonably easy to write and also (I think) to understand: 
```
    const formValues$ = this.props$.map(props => props.formik.values);
    const selectedAccount$ = formValues$.map(x => x.credentials).distinctUntilChanged();
    const selectedRegion$ = formValues$.map(x => x.region).distinctUntilChanged();
    const subnetPurpose$ = formValues$.map(x => x.subnetType).distinctUntilChanged();
```

2) The reactive logic (when source data A or B changes, data C needs to be updated) is separated into small, easily digestible chunks which can be reused.  I.e., when either A) the list of accounts or B) the selected account has changed, the list of regions (C) should be recalculated.  This logic is encapsulated in a reusable observable variable called `regions$` which is then used as input for other reactive data.
```
    const regions$ = Observable.combineLatest(selectedAccount$, accounts$)  // wait for accounts to load and be cached
      .switchMap(([selectedAccount, _accounts]) => AccountService.getRegionsForAccount(selectedAccount))
      .shareReplay(1);
```

3) Side effects are isolated from the mapping logic instead of being colocated.  I think this makes the application state updates easier to follow and debug.
```
    // Update selected zones when the selected region changes
    const regionZones$ = selectedRegion$
      .withLatestFrom(regions$)
      .map(([selectedRegion, regions]) => regions.find(region => region.name === selectedRegion))
      .map(region => (region ? region.availabilityZones : []));
    ...
    regionZones$.takeUntil(this.destroy$).subscribe(regionZones => {
      this.props.formik.setFieldValue('regionZones', regionZones);
    });
```

4) The inputs are predictable and well defined because the logic has been inverted from a push model to a pull model.  It's easy to understand what will cause an observable to update.  I.e., `subnets$` is recomputed when either `selectedAccount$` or `selectedRegion$` change.
```
    const subnets$ = Observable.combineLatest(selectedAccount$, selectedRegion$)
      .switchMap(([credentials, region]) => this.getAvailableSubnets(credentials, region))
      .map(availableSubnets => this.makeSubnetOptions(availableSubnets))
      .shareReplay(1);
```

5) Async and sync are handled in a very similar manner.  When inputs change and some other value needs to be updated, Rx handles the sync or async callback similarly by choosing between `.map()` (sync) and  `.switchMap` or `.mergeMap` (asyc)

###  Cons:

1) Heavy use of RxJS can be confusing to the unfamiliar (and to the familiar)

What is the difference between `switchMap` and `mergeMap`?  When do you use `.shareReplay(1)`?  What is the difference between the similarly named `combineLatest()` and `withLatestFrom()`? There are like a hundred operators... do I have to learn them all?!

2) This adds another mental model which needs to be learned and understood

3) Formik manages form state internally, and this introduces another interaction between Formik and the enclosing React Component.

Example bridge from formik data:
```
    const formValues$ = this.props$.map(props => props.formik.values);
    const selectedAccount$ = formValues$.map(x => x.credentials).distinctUntilChanged();
    const selectedRegion$ = formValues$.map(x => x.region).distinctUntilChanged();
    const subnetPurpose$ = formValues$.map(x => x.subnetType).distinctUntilChanged();

    const appName$ = this.props$.map(props => props.app.name).distinctUntilChanged();
    const stack$ = formValues$.map(x => x.stack).distinctUntilChanged();
    const detail$ = formValues$.map(x => x.detail).distinctUntilChanged();
```

Example bridge to formik data:
```
  subnet$.takeUntil(this.destroy$).subscribe(subnet => {
      // console.log({ subnet });
      this.props.formik.setFieldValue('vpcId', subnet && subnet.vpcIds[0]);
      this.props.formik.setFieldValue('subnetType', subnet && subnet.purpose);
    });

    moniker$.takeUntil(this.destroy$).subscribe(moniker => {
      this.props.formik.setFieldValue('moniker', moniker);
      this.props.formik.setFieldValue('name', moniker.cluster);
    });
```

4) The way I've structured this PR puts a lot of observable logic into the `componentDidMount` method.  I'm not sure how I feel about that.  The only observables placed on the component instance are those things that need to be accessed from component callbacks (such as `onClick={() => this.refresh$.next()}` or `componentDidUnmount() { this.destroy$.next(); }`

### CODE REVIEW

**Comment by erikmunson** [comment]
I _really_ like the way this manifests for broadcasting out to things beyond Formik. The bridge back to `setFieldValue` — while somewhat straightforward in this example — feels a bit like a can of worms.

Maybe part of that perception is because it seems like even with some nice RX constructs on top, we're still introducing bi-directional data flow into Formik values. That part is inherent in the complexity of the the data this form represents, so I imagine it can't really just disappear. But I do worry that if we start doing this more liberally we'll end up with cases of spaghetti RX code that updates Formik in enough discrete places to become hard to follow/change.

Maybe this is totally off base, but it almost feels like we're reaching for something akin to an RX-ish version of Formik's `validate` where you get new Formik state, do async stuff, then return back new state in one unified contract (as opposed to RX as the between separate input/output contracts). Anywho, IMO the pros outweigh the cons here and the RX learning curve is not a huge deal given the level of RX knowledge we already expect (require?) from folks interacting with Deck 👍

File: app/scripts/modules/amazon/src/loadBalancer/configure/common/LoadBalancerLocation.tsx [hunk 1]
```diff
@@ -111,51 +115,102 @@ class LoadBalancerLocationImpl extends React.Component<ILoadBalancerLocationProp
...
-  private loadAccounts(): void {
-    AccountService.listAccounts('aws').then(accounts => {
-      this.setState({ accounts });
-      this.accountUpdated(this.props.formik.values.credentials);
+    const formValues$ = this.props$.map(props => props.formik.values);
+    const selectedAccount$ = formValues$.map(x => x.credentials).distinctUntilChanged();
+    const selectedRegion$ = formValues$.map(x => x.region).distinctUntilChanged();
+    const subnetPurpose$ = formValues$.map(x => x.subnetType).distinctUntilChanged();
+
+    const appName$ = this.props$.map(props => props.app.name).distinctUntilChanged();
```
Comment (anotherchrisberry) [inline app/scripts/modules/amazon/src/loadBalancer/configure/common/LoadBalancerLocation.tsx]: Seems like this would never change and maybe doesn't need to be a stream

File: app/scripts/modules/amazon/src/loadBalancer/configure/common/LoadBalancerLocation.tsx [hunk 2]
```diff
@@ -111,51 +115,102 @@ class LoadBalancerLocationImpl extends React.Component<ILoadBalancerLocationProp
...
+    const allLoadBalancers$ = this.props.app.getDataSource('loadBalancers').data$ as Observable<IAmazonLoadBalancer[]>;
+    const existingLoadBalancerNames$ = Observable.combineLatest(allLoadBalancers$, selectedAccount$, selectedRegion$)
+      .map(([allLoadBalancers, selectedAccount, selectedRegion]) => {
+        return allLoadBalancers
+          .filter(lb => lb.account === selectedAccount && lb.region === selectedRegion)
+          .map(lb => lb.name);
+      })
+      .shareReplay(1);
+
+    const subnets$ = Observable.combineLatest(selectedAccount$, selectedRegion$)
```
Comment (anotherchrisberry) [inline app/scripts/modules/amazon/src/loadBalancer/configure/common/LoadBalancerLocation.tsx]: This is really minor but it's easy to confuse `subnets$` and `subnet$` when reading the code; maybe rename this to `availableSubnets$` or `allSubnets$`?

File: app/scripts/modules/amazon/src/loadBalancer/configure/common/LoadBalancerLocation.tsx [hunk 3]
```diff
@@ -111,51 +115,102 @@ class LoadBalancerLocationImpl extends React.Component<ILoadBalancerLocationProp
...
     }
-  }
 
-  private loadAccounts(): void {
-    AccountService.listAccounts('aws').then(accounts => {
-      this.setState({ accounts });
-      this.accountUpdated(this.props.formik.values.credentials);
+    const formValues$ = this.props$.map(props => props.formik.values);
+    const selectedAccount$ = formValues$.map(x => x.credentials).distinctUntilChanged();
+    const selectedRegion$ = formValues$.map(x => x.region).distinctUntilChanged();
```
Comment (anotherchrisberry) [inline app/scripts/modules/amazon/src/loadBalancer/configure/common/LoadBalancerLocation.tsx]: I really like that these variables are prefixed with `selected` - when I'm reading the code, it makes it easy to see that these streams are from form values. It might be a nice pattern for the others below (`stack$` -> `selectedStack$`, etc).

File: app/scripts/modules/amazon/src/loadBalancer/configure/common/LoadBalancerLocation.tsx [hunk 4]
```diff
@@ -111,51 +115,102 @@ class LoadBalancerLocationImpl extends React.Component<ILoadBalancerLocationProp
...
+    const subnets$ = Observable.combineLatest(selectedAccount$, selectedRegion$)
+      .switchMap(([credentials, region]) => this.getAvailableSubnets(credentials, region))
+      .map(availableSubnets => this.makeSubnetOptions(availableSubnets))
+      .shareReplay(1);
+
+    const subnet$ = Observable.combineLatest(subnets$, subnetPurpose$).map(
+      ([subnets, subnetPurpose]) => subnets && subnets.find(subnet => subnet.purpose === subnetPurpose),
+    );
+
+    // I don't understand why we use subnet.availabilityZones here, but region.availabilityZones below.
```
Comment (anotherchrisberry) [inline app/scripts/modules/amazon/src/loadBalancer/configure/common/LoadBalancerLocation.tsx]: My only guess is this is an artifact of EC2 Classic, where there are no subnets. It probably makes more sense to have the `<AvailabilityZoneSelector>` set the `allZones` prop to `values.regionZones.length > 0 ? values.regionZones : availabilityZones` or something like that.

File: app/scripts/modules/amazon/src/loadBalancer/configure/common/SecurityGroups.tsx [hunk 5]
```diff
@@ -55,121 +54,88 @@ class SecurityGroupsImpl extends React.Component<ISecurityGroupsProps, ISecurity
...
+      const label = FirewallLabels.get('Firewalls');
+      return { securityGroupsRemoved: `${label} removed: ${removed.join(', ')}` };
+    }
+    return {};
   }
 
   private clearRemoved = (): void => {
-    this.props.dirtyCallback(SecurityGroups.label, false);
-    this.setState({ removed: [] });
+    this.setState({ removed: [] }, () => this.props.formik.validateForm());
```
Comment (jrsquared) [inline app/scripts/modules/amazon/src/loadBalancer/configure/common/SecurityGroups.tsx]: neat

**Comment by alanmquach** [comment]
I really like this. Very much. It unravels the web of async dependencies and instead composes them together in observable sequences that are much easier to understand (at least to the trained eye).

While I do share the same concern data flowing from Formik into RX then back into Formik (a la setFieldValue), I think it's mostly paranoia of the possibility for that flow to be cyclical and thus causing an infinite loop.

That, however, is actually an artifact of having form fields that depend on other form fields and we can run into the exact same problem with Promises and callbacks, and if I ever had to untangle a circular dependency, the observable sequences feel much easier to reason about than trying to trace through callback hell.

**Comment by christopherthielen** [comment]
> The bridge back to setFieldValue — while somewhat straightforward in this example — feels a bit like a can of worms. 

> Maybe part of that perception is because ... we're introducing bi-directional data flow into Formik values.  ... I imagine it can't really just disappear. But I do worry that ... we'll end up with spaghetti RX code that updates Formik in enough places to become hard to follow/change.

I somewhat agree with your concern about two way data flow.  I also agree that we can't avoid this scenario because it's inherently part of the form logic.  Even with the previous implementation promises + callbacks, there is still two way data flow.  Would love to hear ideas which could avoid it, but I don't see how.

> almost feels like we're reaching for something akin to an RX-ish version of Formik's validate where you get new Formik state, do async stuff, then return back new state in one unified contract (as opposed to RX as the between separate input/output contracts). 

Sounds like formik + redux-observable?  Code something up and let's see?

File: app/scripts/modules/amazon/src/loadBalancer/configure/common/LoadBalancerLocation.tsx (hunk 1 above)
Comment (christopherthielen) [reply app/scripts/modules/amazon/src/loadBalancer/configure/common/LoadBalancerLocation.tsx]: Yeah, it probably doesn't need to be a stream in this context, but I think of this as defensive coding.  If this component were ever to be used in a different context where the app prop could change, this code would ensure the component still works as expected.

File: app/scripts/modules/amazon/src/loadBalancer/configure/common/LoadBalancerLocation.tsx (hunk 2 above)
Comment (christopherthielen) [reply app/scripts/modules/amazon/src/loadBalancer/configure/common/LoadBalancerLocation.tsx]: I made some changes to the variable names in a new commit to address this, see below

File: app/scripts/modules/amazon/src/loadBalancer/configure/common/LoadBalancerLocation.tsx (hunk 3 above)
Comment (christopherthielen) [reply app/scripts/modules/amazon/src/loadBalancer/configure/common/LoadBalancerLocation.tsx]: I moved the observable variables for forms into a separate object in a separate commit.  Do you like this better?
```js
    const form = {
      account$: formValues$.map(x => x.credentials).distinctUntilChanged(),
      region$: formValues$.map(x => x.region).distinctUntilChanged(),
      subnetPurpose$: formValues$.map(x => x.subnetType).distinctUntilChanged(),
      stack$: formValues$.map(x => x.stack).distinctUntilChanged(),
      detail$: formValues$.map(x => x.detail).distinctUntilChanged(),
    };
```

```js
const accountRegions$ = Observable.combineLatest(form.account$, allAccounts$)
      .switchMap(([currentAccount, _allAccounts]) => AccountService.getRegionsForAccount(currentAccount))
      .shareReplay(1);	      .shareReplay(1);
```

File: app/scripts/modules/amazon/src/loadBalancer/configure/common/LoadBalancerLocation.tsx (hunk 4 above)
Comment (christopherthielen) [reply app/scripts/modules/amazon/src/loadBalancer/configure/common/LoadBalancerLocation.tsx]: >Also, it looks like this got lost in the refactor:

I think that's captured with this (unless I'm not understanding something):

```
  const subnet$ = Observable.combineLatest(regionSubnets$, form.subnetPurpose$).map(
      ([allSubnets, subnetPurpose]) => allSubnets && allSubnets.find(subnet => subnet.purpose === subnetPurpose),
    );
    const availabilityZones$ = subnet$.map(subnet => (subnet ? uniq(subnet.availabilityZones).sort() : []));
```

**Comment by christopherthielen** [comment]
Thanks for the feedback, all.

File: app/scripts/modules/amazon/src/loadBalancer/configure/common/LoadBalancerLocation.tsx (hunk 1 above)
Comment (anotherchrisberry) [reply app/scripts/modules/amazon/src/loadBalancer/configure/common/LoadBalancerLocation.tsx]: 👍

File: app/scripts/modules/amazon/src/loadBalancer/configure/common/LoadBalancerLocation.tsx (hunk 4 above)
Comment (anotherchrisberry) [reply app/scripts/modules/amazon/src/loadBalancer/configure/common/LoadBalancerLocation.tsx]: I see, you got it!
